hledger-lib/Hledger/Read/JournalReader.hs

```diff
@@ -457,14 +457,45 @@ modifiertransactionp = do
   return $ ModifierTransaction valueexpr postings
 
 -- | Parse a periodic transaction
-periodictransactionp :: JournalParser m PeriodicTransaction
+periodictransactionp :: MonadIO m => JournalParser m PeriodicTransaction
 periodictransactionp = do
   char '~' <?> "periodic transaction"
-  lift (skipMany spacenonewline)
-  periodexpr <- lift $ T.strip <$> descriptionp
-  _ <- lift followingcommentp
-  postings <- postingsp Nothing
-  return $ PeriodicTransaction periodexpr postings
+  lift $ skipMany spacenonewline
+  -- XXX periodexprp in Hledger.Data.Dates is a SimpleTextParser, which we can't call directly here.
+  -- Instead, read until two or more spaces and reparse that. More use of two spaces is not ideal.
+  pos <- getPosition
+  periodtxt <- lift singlespacedtextp
+  d <- liftIO getCurrentDay
+  (interval, span) <-
+    case parsePeriodExpr d periodtxt of
+        Right (i,s) -> return (i,s)
+        Left e -> 
+          -- Show an informative error. XXX a bit unidiomatic, check for megaparsec helpers  
+          fail $ -- XXX
```

Comment:
@awjchen if you have any time or advice, this function could use your help. It does reparsing, and I replaced two throwError calls with fail, which might allow too-vague parse errors (not sure).

Reply:
Hey,

I've attempted some changes in https://github.com/awjchen/hledger/commit/fcf95c1db5957f58e7bf2766d79f8b7e505d61a4 to eliminate the "re-parsing". I used the `match` combinator directly on `periodexprp`, which required generalizing the types of the parsers in `Data/Dates.hs` -- I hope this is okay.

I haven't looked at the this date handling stuff before, so I've only tried to make small changes while trying to preserve behaviour. I've tested the changes with `stack install --fast --test && make functest`, but I wouldn't know if I've messed something up.